src/cdk-experimental/menu/menu-key-manager.ts

```diff
@@ -0,0 +1,339 @@
+
+    if (this._isHorizontal()) {
+      this._keyManager.withHorizontalOrientation(_config.directionality.value);
+    } else {
+      this._keyManager.withVerticalOrientation();
+    }
+
+    this._keyManager.change.subscribe(() =>
+      this._subscribeToBubbledEvents(this._keyManager.activeItem!)
+    );
+  }
+
+  /**
+   * If the MenuItem opens a menu, subscribe to that menu's bubbled keyboard events emitter.
+   * @param menuItem the triggering menu item who's attached menu's Keyboard Manager it subscribes
+   * to.
+   */
+  private _subscribeToBubbledEvents(menuItem: CdkMenuItem) {
+    // Stop any existing subscriptions since the focus has changed and any open menu has closed.
+    this._destroyKeyManagerListener.next();
+
+    if (menuItem.hasMenu()) {
+      // If the provided menu item has a menu, when the menu opens subscribe to its
+      // bubbled events emitter allowing this Keyboard Manager to react to events which
+      // occurred in the submenu. For example, a TAB key event should close out the entire
+      // menu tree therefore when TAB is detected in a submenu, its Keyboard Manager emits the
+      // TAB event along with closing its open menu. Next, the listening Keyboard Managers higher
+      // up in the chain can react to this event by closing out its open menus and further
+      // bubbling up the event. In essence mimicking the default event bubbling behavior provided by
+      // nesting elements.
+      menuItem
+        .getMenuTrigger()!
+        .opened.pipe(takeUntil(this._destroyKeyManagerListener))
+        .subscribe(() => {
+          const menu = menuItem.getMenu();
+          if (menu?._getBubbledKeyboardEvents()) {
+            menu
+              ._getBubbledKeyboardEvents()!
+              .pipe(takeUntil(this._destroyKeyManagerListener))
+              .subscribe((keyboardEvent: KeyboardEvent) => this.onKeydown(keyboardEvent, true));
+          }
+        });
+    }
+  }
+
+  /** Place focus on the first menu item. */
+  focusFirstItem() {
+    this._keyManager.setFirstItemActive();
+  }
+
+  /** Place focus on the last menu item. */
+  focusLastItem() {
+    this._keyManager.setLastItemActive();
+  }
+
+  /**
+   * Place focus on the given menu item.
+   * @param item the menu item to focus.
+   */
+  focusItem(item: CdkMenuItem) {
+    this._keyManager.setActiveItem(item);
+  }
+
+  /**
+   * Process the given KeyboardEvent which may result in a Menu being open/closed and/or focus
+   * changing between the managed MenuItems.
+   *
+   * @param event the KeyboardEvent to process
+   * @param bubbled whether or not the KeyboardEvent has bubbled up from a submenu
+   */
+  onKeydown(event: KeyboardEvent, bubbled = false) {
+    let activeItem = this._keyManager.activeItem;
+
+    switch (event.keyCode) {
+      case ENTER:
+      case SPACE:
+        if (activeItem && !bubbled) {
+          if (activeItem.hasMenu()) {
+            this._openActiveMenu();
```

Comment:
With all these methods working on the activeItem, it might make more sense to pass activeItem in as a parameter, instead of grabbing the same item again at the start of each secondary method. That way, the logic flow is clear -- you grab activeItem at the start of onKeydown(), then you do things with it based on the key event. (Furthermore, I don't believe these secondary methods make any other references to 'this', so you could make them file-scoped instead of private object methods.)